chia/pools/pool_puzzles.py

```diff
@@ -0,0 +1,336 @@
+    pubkey = G1Element.from_bytes(pubkey_program.as_atom())
+    return pubkey
+
+
+def uncurry_pool_member_inner_puzzle(inner_puzzle: Program):  # -> Optional[Tuple[Program, Program, Program]]:
+    """
+    Take a puzzle and return `None` if it's not a "pool member" inner puzzle, or
+    a triple of `mod_hash, relative_lock_height, pubkey` if it is.
+    """
+    if not is_pool_member_inner_puzzle(inner_puzzle):
+        return None
+    r = inner_puzzle.uncurry()
+    if r is None:
+        return r
+    inner_f, args = r
+
+    # TARGET_PUZZLE_HASH P2_SINGLETON_PUZZLEHASH OWNER_PUBKEY POOL_REWARD_PREFIX ESCAPE_MODE_PUZZLEHASH
+    target_puzzle_hash, p2_singleton_hash, owner_pubkey, pool_reward_prefix, escape_puzzlehash = list(args.as_iter())
+    # assert p2_singleton_hash == P2_SINGLETON_HASH
+
+    # return target_puzzle_hash, owner_pubkey
+    return inner_f, target_puzzle_hash, p2_singleton_hash, owner_pubkey, pool_reward_prefix, escape_puzzlehash
+
+
+def uncurry_pool_waitingroom_inner_puzzle(inner_puzzle: Program):  # -> Optional[Tuple[Program, Program, Program]]:
+    """
+    Take a puzzle and return `None` if it's not a "pool member" inner puzzle, or
+    a triple of `mod_hash, relative_lock_height, pubkey` if it is.
+    """
+    if not is_pool_member_inner_puzzle(inner_puzzle):
+        return None
+    r = inner_puzzle.uncurry()
+    if r is None:
+        return r
+    inner_f, args = r
+
+    # TARGET_PUZHASH RELATIVE_LOCK_HEIGHT OWNER_PUBKEY P2_SINGLETON_PUZHASH
+    target_puzzle_hash, relative_lock_height, owner_pubkey, p2_singleton_hash = list(args.as_iter())
+    # assert p2_singleton_hash == P2_SINGLETON_HASH
+
+    return target_puzzle_hash, relative_lock_height, owner_pubkey, p2_singleton_hash
+
+
+def get_inner_puzzle_from_puzzle(full_puzzle: Program) -> Optional[Program]:
+    p = Program.from_bytes(bytes(full_puzzle))
+    r = p.uncurry()
+    if r is None:
+        return None
+    inner_f, args = r
+
+    # TODO(adam): fix
+    # if not is_pool_singleton_inner_puzzle(inner_f):
+    #     return None
+    mod_hash, launcher_id, launcher_puzzle_hash, inner_puzzle = list(args.as_iter())
+    return inner_puzzle
+
+
+def solution_to_extra_data(full_spend: CoinSolution) -> Optional[PoolState]:
+    full_solution_ser: SerializedProgram = full_spend.solution
+    full_solution: Program = Program.from_bytes(bytes(full_solution_ser))
+
+    if full_spend.coin.puzzle_hash == SINGLETON_LAUNCHER_HASH:
+        # Launcher spend
+        extra_data = full_solution.rest().rest().first().as_atom()
+        return PoolState.from_bytes(extra_data)
+
+    # Not launcher spend
+    inner_solution: Program = full_solution.rest().rest().first()
+    inner_spend_type: int = inner_solution.first().as_int()
+
+    if inner_spend_type == 0:
+        # Absorb
+        return None
+
+    # Spend which is not absorb, and is not the launcher
+    num_args = len(inner_solution.as_atom_list())
+    assert num_args == 4 or num_args == 5
```

Comment:
Could also just do `assert num_args == 5` in the `else` clause 3 lines below. 
Btw. this line here also equals
`assert num_args in (4, 5)`